### PR TITLE
Fix!: MySQL Timestamp Data Types

### PR DESCRIPTION
**Pull request by eakmanrq**
Correctly maps `TIMESTAMP` -> `DATETIME` and uses `TIMESTAMP` function when creating the timestamp types.

### CODE REVIEW

File: sqlglot/dialects/mysql.py [hunk 1]
```diff
@@ -555,11 +555,22 @@ class Generator(generator.Generator):
...
+                k: v
+                for k, v in generator.Generator.TYPE_MAPPING.items()
+                if k
+                not in {
+                    exp.DataType.Type.MEDIUMTEXT,
+                    exp.DataType.Type.LONGTEXT,
+                    exp.DataType.Type.MEDIUMBLOB,
+                    exp.DataType.Type.LONGBLOB,
+                }
+            },
```
Comment (GeorgeSittas) [inline sqlglot/dialects/mysql.py:569]: This is going to run for every key in `TYPE_MAPPING`, right? Not sure why this change was necessary.

File: sqlglot/dialects/mysql.py [hunk 2]
```diff
@@ -555,11 +555,22 @@ class Generator(generator.Generator):
...
+            exp.DataType.Type.TIMESTAMP: "DATETIME",
+            exp.DataType.Type.TIMESTAMPTZ: "TIMESTAMP",
+            exp.DataType.Type.TIMESTAMPLTZ: "TIMESTAMP",
```
Comment (GeorgeSittas) [inline sqlglot/dialects/mysql.py:572]: I think we can do something similar to what I did [here](https://github.com/tobymao/sqlglot/pull/2172/files#diff-14ef9d8b99f48dcf8f6a4e3bbab7bf55a9d85d2b83d684c3c822f0f94fcd1d07R558) to factor this types out.

File: sqlglot/dialects/mysql.py [hunk 3]
```diff
@@ -599,6 +610,13 @@ def jsonarraycontains_sql(self, expression: exp.JSONArrayContains) -> str:
             return f"{self.sql(expression, 'this')} MEMBER OF({self.sql(expression, 'expression')})"
 
         def cast_sql(self, expression: exp.Cast, safe_prefix: t.Optional[str] = None) -> str:
+            if expression.to.this in {
+                exp.DataType.Type.TIMESTAMP,
+                exp.DataType.Type.TIMESTAMPTZ,
+                exp.DataType.Type.TIMESTAMPLTZ,
```
Comment (GeorgeSittas) [inline sqlglot/dialects/mysql.py:616]: See comment in type mapping w.r.t to drying this section up.

Comment (tobymao) [reply sqlglot/dialects/mysql.py:616]: use the Temporal_types constant

File: sqlglot/dialects/mysql.py (hunk 2 above)
Comment (eakmanrq) [reply sqlglot/dialects/mysql.py:572]: Ah I see you just changed that. I didn't see the code change when I copy/pasted this over. Will fix.

File: sqlglot/dialects/mysql.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/dialects/mysql.py:569]: just do 
```suggestion
TYPE_MAPPING = TYPE_MAPPING.copy()

for from, to in (
  (MEDIUMTEXT, None),
  (TIMESTAMP, "DATETIME")
 : 
 if to:
   TYPE_MAPPING[from] = to
   els:
   TYPE_MAPPING.pop(from)

```

Comment (eakmanrq) [reply sqlglot/dialects/mysql.py]: Replaced with approach used in previous PR

File: sqlglot/dialects/mysql.py (hunk 3 above)
Comment (eakmanrq) [reply sqlglot/dialects/mysql.py:631]: Temporal includes date

Comment (GeorgeSittas) [reply sqlglot/dialects/mysql.py:631]: That covers all temporal types though right, e.g. `TIME`, `DATETIME`, .. This might be valid, but I'm not 100% sure about MySQL's specifics here; if we only care about `TIMESTAMP[TZ|LTZ]` though, we can just do

```python
if expression.to.this in self.TIMESTAMP_TYPE_MAPPING:
```

Comment (tobymao) [reply sqlglot/dialects/mysql.py:631]: yea, just use that constant than, otherwise what's the point of it?

Comment (eakmanrq) [reply sqlglot/dialects/mysql.py]: Excluded DATETIME and then also added Doris support which doesn't seem to have a `TIMESTAMP` func so this changed a bit.

File: sqlglot/dialects/mysql.py [hunk 4]
```diff
@@ -563,9 +563,16 @@ class Generator(generator.Generator):
             exp.DataType.Type.UTINYINT: "TINYINT",
         }
 
+        TIMESTAMP_TYPE_MAPPING = {
+            exp.DataType.Type.TIMESTAMP: "DATETIME",
+            exp.DataType.Type.TIMESTAMPTZ: "TIMESTAMP",
+            exp.DataType.Type.TIMESTAMPLTZ: "TIMESTAMP",
```
Comment (GeorgeSittas) [inline sqlglot/dialects/mysql.py:569]: If this isn't used anymore then we should probably move these inline in `TYPE_MAPPING`

**Review by GeorgeSittas — APPROVED** [review]
A couple last comments but otherwise looks good 👍

File: sqlglot/dialects/mysql.py [hunk 5]
```diff
@@ -618,6 +630,9 @@ def jsonarraycontains_sql(self, expression: exp.JSONArrayContains) -> str:
             return f"{self.sql(expression, 'this')} MEMBER OF({self.sql(expression, 'expression')})"
 
         def cast_sql(self, expression: exp.Cast, safe_prefix: t.Optional[str] = None) -> str:
+            if expression.to.this in self.TIMESTAMP_FUNC_TYPES:
+                return super().func("TIMESTAMP", expression.this)
```
Comment (GeorgeSittas) [inline sqlglot/dialects/mysql.py:634]: No need to do `super` here, `self.func` also works (nit)

File: sqlglot/dialects/mysql.py [hunk 6]
```diff
@@ -592,6 +599,11 @@ class Generator(generator.Generator):
             exp.DataType.Type.VARCHAR: "CHAR",
         }
 
+        TIMESTAMP_FUNC_TYPES = {
```
Comment (tobymao) [inline sqlglot/dialects/mysql.py:602]: move this into DataType

TIMESTAMP_TYPES

you're also missing TIMESTAMP

Comment (tobymao) [reply sqlglot/dialects/mysql.py:602]: nevermind

File: sqlglot/dialects/mysql.py (hunk 5 above)
Comment (tobymao) [inline sqlglot/dialects/mysql.py]:
```suggestion
                return self.func("TIMESTAMP", expression.this)
```
